lib/storage/data/external/GCP/GcpService.js

```diff
@@ -1,141 +1,908 @@
+        // This runs after the SDK decides virtual-hosted vs path-style
+        // and forces path-style for DNS-incompatible buckets
+        this.middlewareStack.add(
+            gcpDnsCompatibilityMiddleware(config),
+            {
+                name: 'gcpDnsCompatibility',
+                step: 'serialize',
+                priority: 'high'
+            }
+        );
 
-Object.assign(GCP.prototype, GcpServiceSetup, {
-    _maxConcurrent: 5,
+        // Add GCP signing middleware
+        this.middlewareStack.add(
+            gcpSigningMiddleware(config),
+            {
+                name: 'gcpSigningMiddleware',
+                step: 'finalizeRequest',
+                priority: 'high'
+            }
+        );
+    }
 
-    // Implemented APIs
-    // Bucket API
-    getBucket(params, callback) {
-        return this.listObjects(params, callback);
-    },
+    /**
+     * Upload an object to GCP (with optional tagging as metadata).
+     * @param {object} params - S3 putObject params
+     * @param {function} callback
+     */
+    putObject(params, callback) {
+        const command = new PutObjectCommand(params);
+        
+        // Capture GCP-specific headers
+        let gcpHeaders = {};
+        command.middlewareStack.add(
+            (next) => async (args) => {
+                const result = await next(args);
+                if (result.response?.headers) {
+                    gcpHeaders = result.response.headers;
+                }
+                return result;
+            },
+            {
+                step: 'deserialize',
+                name: 'captureGcpHeaders',
+            }
+        );
+        
+        return this.send(command)
+            .then(data => {
+                // Add GCP-specific fields
+                const result = { ...data };
+                if (gcpHeaders['x-goog-generation']) {
+                    result.VersionId = gcpHeaders['x-goog-generation'];
+                }
+                return callback && callback(null, result);
+            })
+            .catch(err => callback?.(err));
+    }
+
+    /**
+     * Set object tags (stored as metadata in GCP).
+     * @param {object} params - S3 putObjectTagging params
+     * @param {function} callback
+     */
+    putObjectTagging(params, callback) {
+        if (!params.Tagging || !params.Tagging.TagSet) {
+            return callback(errorInstances.MissingParameter);
+        }
+        const tagRes = processTagSet(params.Tagging.TagSet);
+        if (tagRes instanceof Error) {
+            return callback(tagRes);
+        }
+        return async.waterfall([
+            next => this.headObject({
+                Bucket: params.Bucket,
+                Key: params.Key,
+                VersionId: params.VersionId,
+            }, next),
+            (resObj, next) => {
+                const completeMD = Object.assign({}, resObj.Metadata, tagRes);
+                this.copyObject({
```

Comment:
putObjectTagging should not be implemented with CopyObject : this would actually duplicate the data,
and cause unexpected (service) charges to the customer...